### ISSUE TITLE
Hitting a use after free on Windows with an app that has AUI

### ISSUE THREAD

**Issue by wxtrac** [wxWidgets/wxWidgets#18790]
*Issue migrated from trac ticket # 18790*

**component:** wxMSW | **priority:** normal

#### 2020-06-14 06:46:24: rmpowell77 created the issue

___
I've changed our wxWidgets based project to start using AUI.  Project is at:

https://github.com/calband/calchart

Things work fine on Mac.  But on Windows I see a crash when I close our document.  I've traced it down to what looks like a use after free.

What I see is that when the wxFrame is being destroyed, it first destroys all the members, which includes the member m_winLastFocused.  This is a weakRef to the last focused object.

Next what happens is that the children are then destroyed.  I see this:

```
#!cpp
wxTopLevelWindowMSW::~wxTopLevelWindowMSW()
{
    delete m_menuSystem;

    SendDestroyEvent();

    // after destroying an owned window, Windows activates the next top level
    // window in Z order but it may be different from our owner (to reproduce
    // this simply Alt-TAB to another application and back before closing the
    // owned frame) whereas we always want to yield activation to our parent
    if ( HasFlag(wxFRAME_FLOAT_ON_PARENT) )
    {
        wxWindow *parent = GetParent();
        if ( parent )
        {
            ::BringWindowToTop(GetHwndOf(parent));
        }
    }
}
```

So what happens is that the child calls out to the parent to activate it, but the parent has already been partially deleted, specifically that the m_winLastFocused wxWeakRef has been delete.  This cause the item to be used after it's had it's dtor called.

I found I can resolve this issue by change the OnActivate member function in wxTopLevelWindowMSW

```
#!cpp
void wxTopLevelWindowMSW::OnActivate(wxActivateEvent& event)
{
    if ( event.GetActive() )
    {
```

If I insert a check to see if the object is being destroyed, this can avoid this problem:

```
#!cpp
void wxTopLevelWindowMSW::OnActivate(wxActivateEvent& event)
{
    if (IsBeingDeleted()) {
         event.Skip();
         return;
    }
    if ( event.GetActive() )
    {
```

I *think* this helps resolve the issue, but probably what it means is that the focus may not be given to the right document after we've closed it down.  But it's better than crashing.

**Comment by vadz** [comment]
I think this may have been accidentally fixed by dda052d38ab02951d0c3bc65adff3d14f8f40c38 (see #18535), so it should be closed until it can still be reproduced. If the OP (@rmpowell77?) is still there, please check if it can be with the current versions. TIA!

**Comment by rmpowell77** [comment]
I'll have to dig up the windows version and see if I can still reproduce it.
